python/psim/plugins/snapshot.py

```diff
@@ -0,0 +1,65 @@
+"""Set of plugins used to generate a 'snapshot' of the simulation state.
+"""
+
+from psim.plugins import Plugin
+
+import logging
+
+log = logging.getLogger(__name__)
+
+
+class Snapshot(Plugin):
+    """Captures a snapshot of the simulation state upon termination and saves it
+    to the specified file.
+    """
+    def __init__(self, snapshot=None):
+        super(Snapshot, self).__init__()
+
+        self._snapshot = snapshot
+
+    def arguments(self, parser):
+        super(Snapshot, self).arguments(parser)
+
+        parser.add_argument(
+            '--snapshot', type=str, help='specifies the output file for the ' +
+            'simulation snapshot upon simulation termination.'
+        )
+
+    def initialize(self, sim, args):
+        super(Snapshot, self).initialize(sim, args)
+
+        if not args.snapshot:
+            if not self._snapshot:
+                log.warning('No snapshot file specified; snapshot feature disabled.')
+            else:
+                log.warning('No snapshot file specified; default to "%s"', self._snapshot)
+
+            return
+        
+        self._snapshot = args.snapshot
+        log.info('Saving simulation snapshot to "%s" upon simulation termination.', self._snapshot)
+
+    def cleanup(self, sim):
+        super(Snapshot, self).cleanup(sim)
+
+        if not self._snapshot:
+            return
+
+        _fields = [
```

Comment:
Would there be any reason for a list like this to be passed in as, say, a filename via a psim command line option? Like do we expect this set of parameters to change? If so, might be worth putting in the work now to have it work?

Reply:
Given that we'd only be tracking truth data here, no I don't foresee this changing too much. Also, the feature is only used to generate new test cases so from the perspective I don't think the extra work is really worth it atm.